### PR TITLE
Add member tab and add member invitation

### LINKED ISSUE #117: Implement: Adding project member
I think that it's better for project member management in a separate tab from the `Edit` tab. Let's call that tab `Member`.

A member can invite another member via a form. A member can leave a project with a button.

### PR DESCRIPTION
**Pull request by Huy-Ngo**
Draft for now, since I intend to resolve #117 

I'm not sure if I should delete the member list from info so I kept it as a separate commit till I hear your opinion.

Should we put an invitation notification at home page?

### CODE REVIEW

File: src/acanban/project.py [hunk 1]
```diff
@@ -86,3 +86,57 @@ async def edit(uuid: str) -> ResponseReturnValue:
...
+    """Add a member to the project."""
+    project_query = r.table('projects').get(uuid)
+    async with current_app.db_pool.connection() as connection:
+        try:
+            project = await project_query.pluck(*BASIC_FIELDS).run(connection)
+        except ReqlNonExistenceError:
+            raise NotFound
+        supervisors = project['supervisors']
+        students = project['students']
+        # FIX ME: checking student and supervisors not working
```
Comment (Huy-Ngo) [inline src/acanban/project.py:123]: The check is in the `if` block 9 lines below but ok

**Comment by Huy-Ngo** [comment]
Please pull rebase from main, so that the stubs are included

Current lint fail is:

```
src/acanban/project.py:136: error: "RethinkDB" has no attribute "row"
src/acanban/project.py:141: error: "RethinkDB" has no attribute "row"
```

File: src/acanban/project.py [hunk 2]
```diff
@@ -155,3 +155,57 @@ async def report_eval(uuid: str) -> ResponseReturnValue:
...
+            if project['id'] in user['projects']:
+                await flash('That user is already a member of the project')
+                return redirect(f'/p/{uuid}/members')
+            updated_projects = r.row['projects'].append(uuid)
+            await r.table('users').get(new_name).update(
+                {'projects': updated_projects}).run(connection)
+            mem_field = f'{user["role"]}s'
+            await project_query.update(
+                {mem_field: r.row[mem_field].append(new_name)}).run(connection)
+    return redirect(f'/p/{uuid}/members')
```
Comment (McSinyx) [inline src/acanban/project.py:211]: The conditionals here are hard to follow.  If you want to deliver a message *and* short-circuit the execution, use an exception instead of the flash-return pair.  Plus some exceptions give you different status code, which makes testing easier.

File: src/acanban/project.py [hunk 3]
```diff
@@ -155,3 +155,57 @@ async def report_eval(uuid: str) -> ResponseReturnValue:
...
+        try:
+            project = await project_query.pluck(*BASIC_FIELDS).run(connection)
+        except ReqlNonExistenceError:
+            raise NotFound
+        supervisors = project['supervisors']
+        students = project['students']
+        if (current_user.key not in supervisors
+                and current_user.key not in students
+                and current_user.role != 'assistant'):
+            raise Unauthorized
```
Comment (McSinyx) [inline src/acanban/project.py:175]: Use `pluck` defined above.

File: src/acanban/project.py [hunk 4]
```diff
@@ -155,3 +155,57 @@ async def report_eval(uuid: str) -> ResponseReturnValue:
...
+        try:
+            project = await project_query.pluck(*BASIC_FIELDS).run(connection)
+        except ReqlNonExistenceError:
+            raise NotFound
+        supervisors = project['supervisors']
+        students = project['students']
+        # FIX ME: checking student and supervisors not working
+        if (current_user.key not in supervisors
+                and current_user.key not in students):
+            raise Unauthorized
```
Comment (McSinyx) [inline src/acanban/project.py:195]: Use `pluck` defined above.

File: src/acanban/templates/project-info.html [hunk 5]
```diff
@@ -4,10 +4,6 @@
 {% endblock %}
 {% block tabcontent %}
 <div class=col-wide>
-  <p><b>Supervisors:</b> {{ project.supervisors | map('userlink')
-                            | join(', ') | safe }}</p>
-  <p><b>Students:</b> {{ project.students | map('userlink')
-                         | join(', ') | safe }}</p>
```
Comment (McSinyx) [inline src/acanban/templates/project-info.html:10]: With this removed, make sure you edit BASIC_FIELDS to not contain information not used by `info` and `edit`.

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: src/acanban/project.py (hunk 4 above)
Comment (9a24f0) [reply src/acanban/project.py]: Addressed at 537fac5e0fb1b5044d0e36abe68b456fcccfa562

File: src/acanban/project.py (hunk 3 above)
Comment (9a24f0) [reply src/acanban/project.py]: Addressed at 537fac5e0fb1b5044d0e36abe68b456fcccfa562

File: src/acanban/project.py [hunk 6]
```diff
@@ -155,3 +155,35 @@ async def report_eval(uuid: str) -> ResponseReturnValue:
...
+    return await render_template('project-member-list.html', project=project)
+
+
+@blueprint.route('/<uuid>/invite', methods=['POST'])
+@login_required
+async def invite_member(uuid: str) -> ResponseReturnValue:
+    """Add a member to the project."""
+    project = await pluck(uuid, BASIC_FIELDS)
+    form = await request.form
+    new_name = form['new-user']
```
Comment (McSinyx) [inline src/acanban/project.py:174]: LOL

File: src/acanban/project.py [hunk 7]
```diff
@@ -155,3 +155,35 @@ async def report_eval(uuid: str) -> ResponseReturnValue:
...
+    async with current_app.db_pool.connection() as conn:
+        user = await r.table('users').get(new_name).run(conn)
```
Comment (McSinyx) [inline src/acanban/project.py:176]: To minimize the I/O, you might want to `.pluck()` here.

File: src/acanban/project.py (hunk 2 above)
Comment (McSinyx) [reply src/acanban/project.py]: This is still hard to follow.  Once again I suggest short-circuiting to unnest the conditionals.  Exceptions doesn't seem to be the right thing to do here, so probably try to return the rendered page with the error directly.  That will give you 200 on user input error, 302 on success (btw there's `request.referrer`) and 4xx on other user errors.

File: src/acanban/project.py (hunk 7 above)
Comment (McSinyx) [reply src/acanban/project.py:193]: My mistake, `user['projects']` is used below.

**Review by McSinyx — APPROVED** [review]
The tests better be more assertive, but from the higher-level view the implementation is fine I guess.

**Comment by Huy-Ngo** [comment]
You can add tests later (make an issue for reminder if necessary).
I think all existing tests are quite lousy anyway.
